Extract Module to Library (#4274)
facebook/react-native-website: docs/the-new-architecture/create-module-library.md

# Create a Library for Your Module

> React Native has a florid ecosystem of libraries to solve common problems. We collect React Native libraries in the [`reactnative.directory`](https://reactnative.directory) website, and this is a great resource to bookmark for every React Native developer.

Reviewer: ```suggestion
React Native has a florid ecosystem of libraries to solve common problems. We collect React Native libraries in the [reactnative.directory](https://reactnative.directory) website, and this is a great resource to bookmark for every React Native developer.
```

Sometimes, you might be working on a module that is worth extracting in a separate library for code reuse. This can be a library that you want to reuse in all your apps, a library that you want to distribute to the ecosystem as an open source component, or even a library you'd like to sell.

In this guide, you'll learn:

- how to extract a module into a library
- how to distribute the library using NPM

## Extract the Module into a Library

> The React Native community produced an amazing tool to scaffold a new library: the [`create-react-native-library`](https://callstack.github.io/react-native-builder-bob/create) tool. This tool sets up a new library with all the boilerplate code that is needed: all the configuration files and all files required by the various platforms. It also comes with a nice interactive menu to guide you through the creation of the library.

Reviewer: ```suggestion
You can use the [`create-react-native-library`](https://callstack.github.io/react-native-builder-bob/create) tool to create a new library. This tool sets up a new library with all the boilerplate code that is needed: all the configuration files and all files required by the various platforms. It also comes with a nice interactive menu to guide you through the creation of the library.
```

That's it! You have moved all the required code out of your app and in a separate library.

## Testing your Library

The `create-react-native-library` comes with an useful example application that is already configured to work properly with the library. This is a great way to test it!

If you look at the `example` folder, you can find the same structure of a new React Native application that you can create from the [`react-native-community/template`](https://github.com/react-native-community/template).

To test your library:

1. Navigate to the `example` folder.
2. Run `yarn install` to install all the dependencies.
3. For iOS only, you need to install CocoaPods: `cd ios && pod install`.
4. Build and run Android with `yarn android` from the `example` folder.
5. Build and run iOS with `yarn ios` from the `example` folder.

> 

Author: I wanted to add a paragraph on the Local Module setup, but it is currently broken. Metro can't follow the symlinks that yarn generates and it can't pick up the JS files from the other folder.

Probably there is a way to make it work by tinkering with the `metro.config.js`, but I don't think we should go to that level of details in these guides, so I decided to leave that setup out for simplicity. There is a way to test the library anyway.

cc. @huntie if you know how to handle this case in a simple way.
